File: src/main/java/etlmail/front/gui/choosetemplate/FileDocument.java
package etlmail.front.gui.choosetemplate;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import javax.swing.event.DocumentListener;
import javax.swing.event.UndoableEditListener;
import javax.swing.text.*;

import etlmail.front.gui.helper.DocumentAdapter;
import etlmail.front.gui.helper.ModelUtils;

public class FileDocument implements Document {
    private final Document wrapped;
    private final List<FilenameListener> listeners = new ArrayList<FilenameListener>();

    public FileDocument(Document wrapped) {
	this.wrapped = wrapped;
	wrapped.addDocumentListener(new DocumentAdapter() {
	    @Override
	    protected void update(String newText) {
		for (final FilenameListener listener : listeners) {
		    listener.update(new File(newText));
		}
	    }
	});
    }

    public FileDocument() {
	this(new DefaultStyledDocument());
    }

    public File getFile() {
	return new File(ModelUtils.getText(wrapped));
    }

    public void setFile(File file) {
	ModelUtils.setText(wrapped, file.getPath());
    }

    public void addFilenameListener(FilenameListener listener) {
	listeners.add(listener);
    }

    @Override
    public void addDocumentListener(DocumentListener arg0) {
	wrapped.addDocumentListener(arg0);
    }

    @Override
    public void addUndoableEditListener(UndoableEditListener arg0) {
	wrapped.addUndoableEditListener(arg0);
    }

    @Override
    public Position createPosition(int arg0) throws BadLocationException {
	return wrapped.createPosition(arg0);
    }

    @Override
    public Element getDefaultRootElement() {
	return wrapped.getDefaultRootElement();
    }

    @Override
    public Position getEndPosition() {
	return wrapped.getEndPosition();
    }

    @Override
    public int getLength() {
	return wrapped.getLength();
    }

    @Override
    public Object getProperty(Object arg0) {
	return wrapped.getProperty(arg0);
    }

    @Override
    public Element[] getRootElements() {
	return wrapped.getRootElements();
    }

    @Override
    public Position getStartPosition() {
	return wrapped.getStartPosition();
    }

    @Override
    public String getText(int arg0, int arg1) throws BadLocationException {
	return wrapped.getText(arg0, arg1);
    }

    @Override
    public void getText(int arg0, int arg1, Segment arg2) throws BadLocationException {
	wrapped.getText(arg0, arg1, arg2);
    }

    @Override
    public void insertString(int arg0, String arg1, AttributeSet arg2) throws BadLocationException {
	wrapped.insertString(arg0, arg1, arg2);
    }

    @Override
    public void putProperty(Object arg0, Object arg1) {
	wrapped.putProperty(arg0, arg1);
    }

    @Override
    public void remove(int arg0, int arg1) throws BadLocationException {
	wrapped.remove(arg0, arg1);
    }

    @Override
    public void removeDocumentListener(DocumentListener arg0) {
	wrapped.removeDocumentListener(arg0);
    }

    @Override
    public void removeUndoableEditListener(UndoableEditListener arg0) {
	wrapped.removeUndoableEditListener(arg0);
    }

    @Override
    public void render(Runnable arg0) {
	wrapped.render(arg0);
    }
}

File: src/main/scala/etlmail/front/gui/choosetemplate/FileDocument.scala
package etlmail.front.gui.choosetemplate

import java.io.File
import java.util.ArrayList
import java.util.List

import javax.swing.event.DocumentListener
import javax.swing.event.UndoableEditListener
import javax.swing.text._

import etlmail.front.gui.helper.DocumentAdapter
import etlmail.front.gui.helper.ModelUtils

import scala.collection.JavaConversions._

class FileDocument(wrapped: Document) extends Document {
  private val listeners = new ArrayList[FilenameListener]

  wrapped.addDocumentListener(new DocumentAdapter {
    protected def update(newText: String) {
      for (listener <- listeners) {
        listener.update(new File(newText))
      }
    }
  })

  def this() {
    this(new DefaultStyledDocument)
  }

  def getFile(): File = new File(ModelUtils.getText(wrapped))

  def setFile(file: File) {
    ModelUtils.setText(wrapped, file.getPath)
  }

  def addFilenameListener(listener: FilenameListener) {
    listeners.add(listener)
  }

  def addDocumentListener(listener: DocumentListener) {
    wrapped.addDocumentListener(listener)
  }

  def addUndoableEditListener(listener: UndoableEditListener) {
    wrapped.addUndoableEditListener(listener)
  }

  def createPosition(arg0: Int): Position = wrapped.createPosition(arg0)

  def getDefaultRootElement(): Element = wrapped.getDefaultRootElement

  def getEndPosition(): Position = wrapped.getEndPosition

  def getLength(): Int = wrapped.getLength

  def getProperty(arg0: Any): Object = wrapped.getProperty(arg0)

  def getRootElements(): Array[Element] = wrapped.getRootElements

  def getStartPosition(): Position = wrapped.getStartPosition

  def getText(arg0: Int, arg1: Int): String = wrapped.getText(arg0, arg1)

  def getText(arg0: Int, arg1: Int, arg2: Segment) {
    wrapped.getText(arg0, arg1, arg2)
  }

  def insertString(arg0: Int, arg1: String, arg2: AttributeSet) {
    wrapped.insertString(arg0, arg1, arg2)
  }

  def putProperty(arg0: Any, arg1: Any) {
    wrapped.putProperty(arg0, arg1)
  }

  def remove(arg0: Int, arg1: Int) {
    wrapped.remove(arg0, arg1)
  }

  def removeDocumentListener(arg0: DocumentListener) {
    wrapped.removeDocumentListener(arg0)
  }

  def removeUndoableEditListener(arg0: UndoableEditListener) {
    wrapped.removeUndoableEditListener(arg0)
  }

  def render(arg0: Runnable) {
    wrapped.render(arg0)
  }
}
